### DISCUSSION TITLE
Video transcoding is not working on TrueNAS installations

### DISCUSSION THREAD

**Discussion by immanuelsmith** [photoprism/photoprism#3574] [Q&A]
#### 1. What is not working as documented?

Video playback of iPhone HEVC video is not working. 

#### 2. How can we reproduce it?

Add an originals directory containing an iPhone video and index it. Try to open it from the videos folder or by manually browsing to its location. It will fail to play and instead show a 1-second video of CRT static

#### 3. What behavior do you expect?

The video should play. It plays regular MP4's

#### 4. What could be the cause of your problem?

I think it has something to do with transcoding. I have tried to turn off the FFPMEG transcoder and it has no effect.

#### 5. Can you provide us with example files for testing, error logs, or screenshots?

I don't have any I can share as they are all personal videos, but any regular iPhone HEVC video will cause the issue (has to be .mov format, NOT slo-mo or any other type of video requiring a settings change).

Log with FFMPEG disabled: 
2023-07-25 07:13:31 ERRO video: transcoding '******/05404_IMG_1747.MOV' failed

Log with FFMPEG enabled: 
 2023-07-25 06:43:09 ERRO video: transcoding '****/01544_3A325C8B-B91A-40D6-B3BB-5DAC880D4CE6.MOV' failed
2023-07-25 06:43:09 WARN libx264: failed transcoding ****/01544_3A325C8B-B91A-40D6-B3BB-5DAC880D4CE6.MOV [145.628904ms]
2023-07-25 06:43:09 INFO libx264: transcoding ****/01544_3A325C8B-B91A-40D6-B3BB-5DAC880D4CE6.MOV to avc

#### 6. Which software versions do you use?

Photoprism Build [221118-e58fee0]
Current community build for truenas as of 7/25

From Truenas -
Version:
g20221118
Plugin:
photoprism
Release:
12.4-RELEASE-p3

Stock MariaDB database, no changes made to anything.
#### 7. On what kind of device is PhotoPrism installed?

NAS with a AMD Ryzen 5 3600 6-Core Processor, 16GB of ECC memory, 3x6TB drives in a ZFS pool (Raid 5). It has an Nvidia card for display output but I don't think it is supported by truenas core for any sort of video acceleration

#### 8. Do you use a Reverse Proxy, Firewall, VPN, or CDN?
I don't think so.

**Comment by graciousgrey** [comment]
If videos are to be played by a browser that does not natively support the video format, they must be transcoded. This means that ffmpeg must be enabled.

See also:
 - https://docs.photoprism.app/user-guide/organize/video/#transcoding

In your case the video transcoding does not seem to work.

We've heard similar reports of problems with videos from other TrueNAS users. Unsure what could be causing this. As playing videos is not an issue with other installation methods it's most likely a config issue or a third party integration that doesn't include all required dependencies. 

Is the third party app you are using provided by the community or by the TrueNAS team? Is there any documentation available?

**Reply by stavros-k** [reply]
The "App" is a helm chart using the official docker container `photoprism/photoprism:230719`
Assuming the container have all the needed dependencies, is there any other dependency needed on the host that is not normally installed?

**Reply by lastzero** [reply]
@stavros-k this was for you, posted my reply under the wrong comment: https://github.com/photoprism/photoprism/discussions/3574#discussioncomment-6566001

**Reply by immanuelsmith** [reply]
Is that a docker image I referenced? The truenas installation is truenas core so it should be a bsd jail package

**Reply by lastzero** [reply]
In that case, it makes sense that ffmpeg could be missing!?

**Reply by stavros-k** [reply]
@immanuelsmith Yes, both the link you posted (https://github.com/truenas/charts/tree/master/charts/photoprism) and the issue created in the repo (https://github.com/truenas/charts/issues/1404) are using a container and its only available on TrueNAS SCALE.

**Reply by immanuelsmith** [reply]
Ok well I have no idea how to find the truenas core repo. That being said people have run into the same issue with scale. When I checked the shell on my install FFmpeg is indeed installed so not sure what went wrong or why it is trying to trigger it for transcoding when I explicitly turn that option off

**Reply by lastzero** [reply]
If you turn FFmpeg off then no still images can be created from the video to be used as thumbnails and playing videos with containers or codecs your browser pretends not to support cannot be played. Note that the frontend tests support for common formats via JS to avoid unnecessary transcoding.

**Comment by graciousgrey** [comment]
In order for us to help resolve issues with the TrueNAS installations, it would be great if you could send us links to the installation instructions / app store listings / documentation that you used to install PhotoPrism. Also, it would be helpful to know how to contact the maintainers of these third party apps. Thank you very much!

**Comment by immanuelsmith** [comment]
The photoprism install is an community maintained plugin for Truenas, there are no instructions as it is a one click install. The source of it is here: https://github.com/truenas/charts/tree/master/charts/photoprism . I will crosspost this as an issue there.

**Reply by lastzero** [reply]
That's quite a lot of files, so I couldn't check them all. In our experience, some of the dependencies in our Docker image might stop working if basic system variables like `PATH` were changed in the container configuration or if, for example, kernel security modules restricted certain privileges or kernel calls:

- https://docs.photoprism.app/getting-started/troubleshooting/docker/#kernel-security

From what I remember, Darktable was particularly affected by this. I can imagine that FFmpeg might also use special calls for faster transcoding - although this is just a wild guess.

Problems could theoretically also occur if the container is running under certain user IDs. See our documentation for the UID ranges that should be safe to use:

- https://docs.photoprism.app/getting-started/config-options/#docker-image

**Reply by Toadlips72** [reply]
Hey there @lastzero, @graciousgrey , and @immanuelsmith I think there is a very good chance that this issue is resolved!  See: [https://github.com/photoprism/photoprism/issues/3659#event-10455019855](url)

As it turns out, I had been running a v12.4 jail on TrueNAS.  I had followed the guide at [https://www.truenas.com/community/threads/install-guide-photoprism-in-a-jail-on-truenas-core-13-0.106992/](url) and had used the same prebuilt version g20221118 as @immanuelsmith from the photoprism-freebsd-port repository.

@immanuelsmith, if you want to be absolutely certain, you could turn on the extra logging, and you'll probably see the same error I was seeing:

```
[vost#0:0/libx264 @ 0x808013200] One of -r/-fpsmax was specified together a non-CFR -vsync/-fps_mode. This is contradictory.
```

In any case, all of my issues were resolved by creating a 13.2 jail, setting my repository to "daily" rather than quarterly (in order to get the very latest packages for gettext and ffmpeg), and compiling using the makefile at [https://github.com/huo-ju/photoprism-freebsd-port](url).

As a freebsd neophyte, I found the build process to be quite _interesting_, and it took a lot longer than I ever thought possible, but it ultimately worked!  Thank you, @lastzero and @graciousgrey for your help!!!

**Comment by tschettervictor** [comment]
This is a script I built for PhotoPrism that some folks over at the TrueNAS forums are having the same issue with.
https://github.com/tschettervictor/truenas-iocage-photoprism

I was not able to reproduce the issue though. I uploaded a HEVC video and it seemed to play ok on my end. Using brave browser.

**Comment by immanuelsmith** [comment]
After much investigating I found the cause of this issue. My GPU is no longer supported with the Nvidia driver in EE. There is no GUI notification of it not being supported, I was only able to figure out this when I realized it should have appeared in the app config (which it didn't), and nvidia-smi was not able to see it.
